util/syncpacket/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+#![recursion_limit="128"]
+
+// Needs to be "extern crate" even in rust 2018:
+// https://blog.rust-lang.org/2018/12/21/Procedural-Macros-in-Rust-2018.html
+extern crate proc_macro;
+
+use proc_macro2::Ident;
+use quote::quote;
+use syn::{Attribute, Data, DeriveInput, Expr, Meta, MetaList, NestedMeta, Result, Variant};
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
+
+/// Given a list of arguments to the "protocol" attribute, check
+/// that there is only one argument and return it
+fn parse_protocol_arguments(args: &MetaList) -> Result<&Ident> {
+	if args.nested.len() != 1 {
+		return Err(syn::Error::new_spanned(args, "protocol attribute should have exactly one argument"));
+	}
+
+	// We have exactly one argument, should not panic
+	match args.nested.first().expect("protocol attribute without value").value() {
+		// Meta argument
+		NestedMeta::Meta(meta) => match meta {
+			Meta::Word(ident) => Ok(&ident),
+			_ => return Err(syn::Error::new_spanned(meta, "nested arguments to protocol are not allowed")),
+		},
+		// Quoted string
+		a @ _ => return Err(syn::Error::new_spanned(a, "protocol argument must be an unquoted identifier"))
+	}
+}
+
+/// Helper function to parse arguments to the protocol attribute.
+/// Syntax should be #[protocol(P)] PacketName = 0xNN,
+fn parse_protocol_attribute(input: &Attribute) -> Result<Ident> {
+	let argument = match input.parse_meta() {
+		Ok(arg) => arg,
+		Err(err) => return Err(err)
+	};
+
+	// Arguments to invocation attributes are delivered as a list
+	match argument {
+		Meta::Word(_) => Err(syn::Error::new_spanned(input, "protocol attribute without argument")),
+		Meta::List(args) => parse_protocol_arguments(&args).map(|ok| ok.clone()),
```

Comment:
Put the happy path first?

Reply:
Should this check that there's a single argument in the list as well?

Reply:
This is checked in `parse_protocol_arguments`, do you think here would be a better place?